In [1]:
import pandas as pd

df = pd.read_csv('train.csv', header=0, names=['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', \
                                     'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', \
                                     'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', \
                                     'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', \
                                     'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond', \
                                     'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', \
                                     'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', \
                                     'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', \
                                     'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', \
                                     'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', \
                                     'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', \
                                     'GarageArea', 'GarageQual', 'GarageCond', 'PavedDrive', 'WoodDeckSF', 'OpenPorchSF', \
                                     'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PoolQC', 'Fence', 'MiscFeature', \
                                     'MiscVal', 'MoSold', 'YrSold', 'SaleType', 'SaleCondition', 'SalePrice'])

In [2]:
df.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


In [3]:
col_names = ['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', \
            'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', \
            'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', \
            'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', \
            'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond', \
            'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', \
            'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', \
            'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', \
            'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', \
            'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', \
            'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', \
            'GarageArea', 'GarageQual', 'GarageCond', 'PavedDrive', 'WoodDeckSF', 'OpenPorchSF', \
            'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PoolQC', 'Fence', 'MiscFeature', \
            'MiscVal', 'MoSold', 'YrSold', 'SaleType', 'SaleCondition', 'SalePrice']

In [4]:
missing_percentage = df.isna().sum() * 100 / len(df)
print(missing_percentage[missing_percentage > 0])

LotFrontage     17.739726
Alley           93.767123
MasVnrType      59.726027
MasVnrArea       0.547945
BsmtQual         2.534247
BsmtCond         2.534247
BsmtExposure     2.602740
BsmtFinType1     2.534247
BsmtFinType2     2.602740
Electrical       0.068493
FireplaceQu     47.260274
GarageType       5.547945
GarageYrBlt      5.547945
GarageFinish     5.547945
GarageQual       5.547945
GarageCond       5.547945
PoolQC          99.520548
Fence           80.753425
MiscFeature     96.301370
dtype: float64


In [5]:
df=df.drop(['Fence', 'MiscFeature', 'PoolQC', 'MasVnrType', 'Alley'], axis=1)

In [6]:
missing_percentage = df.isna().sum() * 100 / len(df)
print(missing_percentage[missing_percentage > 0])

LotFrontage     17.739726
MasVnrArea       0.547945
BsmtQual         2.534247
BsmtCond         2.534247
BsmtExposure     2.602740
BsmtFinType1     2.534247
BsmtFinType2     2.602740
Electrical       0.068493
FireplaceQu     47.260274
GarageType       5.547945
GarageYrBlt      5.547945
GarageFinish     5.547945
GarageQual       5.547945
GarageCond       5.547945
dtype: float64


In [7]:
df['SalePrice'].describe()

count      1460.000000
mean     180921.195890
std       79442.502883
min       34900.000000
25%      129975.000000
50%      163000.000000
75%      214000.000000
max      755000.000000
Name: SalePrice, dtype: float64

In [8]:
import pandas as pd
import math
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression, GammaRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, f1_score, recall_score, mean_squared_error
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
import os
os.environ["LOKY_MAX_CPU_COUNT"] = "4"


In [9]:
X = df.drop('SalePrice', axis=1)
X['Electrical'] = X['Electrical'].bfill()
y = df['SalePrice']

In [10]:
# Define feature groups - numeric, ordinal and nominal(categorical)

numeric_cols = X.select_dtypes(include='number').columns.tolist()

ordinal_cols = [
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'HeatingQC',
    'KitchenQual',
    'FireplaceQu',
    'GarageQual',
    'GarageCond'
]

categorical_cols = X.select_dtypes(include=['object', 'string', 'category']).columns.tolist()
nominal_cols = [col for col in categorical_cols if col not in ordinal_cols]

# 2. Define the order for each ordinal column
quality_order = ['Po', 'Fa', 'TA', 'Gd', 'Ex']

ordinal_categories = [
    # ExterQual
    quality_order, 
    # ExterCond
    quality_order, 
    # BsmtQual
    quality_order,
    # BsmtCond
    quality_order,
    # HeatingQC
    quality_order,
    # KitchenQual
    quality_order,
    # FireplaceQu
    quality_order, 
    # GarageQual
    quality_order,
    # GarageCond
    quality_order  
]

numeric_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

ordinal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(categories=ordinal_categories,
                               handle_unknown='use_encoded_value',
                               unknown_value=-1))
])

nominal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer([
    ('num', numeric_pipeline, numeric_cols),
    ('ord', ordinal_pipeline, ordinal_cols),
    ('nom', nominal_pipeline, nominal_cols)
])

lr_no_pca = Pipeline([
    ('preprocessor', preprocessor),
    ('regressor', GammaRegressor(max_iter=10000))
    #('regressor', LinearRegression())
])

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
#X_train, X_test, y_train, y_test = train_test_split(X, y_log, test_size=0.2, random_state=42)

In [12]:
lr_no_pca.fit(X_train, y_train)

y_pred_no_pca = lr_no_pca.predict(X_test)
y_pred_no_pca_train = lr_no_pca.predict(X_train)
y_pred_no_pca_test = lr_no_pca.predict(X_test)
rmse_train = math.sqrt(mean_squared_error(y_train, y_pred_no_pca_train))
rmse_test = math.sqrt(mean_squared_error(y_test, y_pred_no_pca_test))
print(f'Training set RMSE: {rmse_train}')
print(f'Test set RMSE: {rmse_test}')
gap = rmse_train - rmse_test
print("RMSE gap:", gap)
print("Gap ratio:", rmse_test / rmse_train)

Training set RMSE: 44754.99032932205
Test set RMSE: 31322.115102879
RMSE gap: 13432.875226443048
Gap ratio: 0.699857487900243


In [13]:
import pandas as pd
import math
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression, GammaRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, f1_score, recall_score, mean_squared_error
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer


X = df.drop('SalePrice', axis=1)
X['Electrical'] = X['Electrical'].bfill()
y = df['SalePrice']
# Define feature groups - numeric, ordinal and nominal(categorical)

numeric_cols = X.select_dtypes(include='number').columns.tolist()

ordinal_cols = [
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'HeatingQC',
    'KitchenQual',
    'FireplaceQu',
    'GarageQual',
    'GarageCond'
]

categorical_cols = X.select_dtypes(include=['object', 'string', 'category']).columns.tolist()
nominal_cols = [col for col in categorical_cols if col not in ordinal_cols]

# 2. Define the order for each ordinal column
quality_order = ['Po', 'Fa', 'TA', 'Gd', 'Ex']

ordinal_categories = [
    # ExterQual
    quality_order, 
    # ExterCond
    quality_order, 
    # BsmtQual
    quality_order,
    # BsmtCond
    quality_order,
    # HeatingQC
    quality_order,
    # KitchenQual
    quality_order,
    # FireplaceQu
    quality_order, 
    # GarageQual
    quality_order,
    # GarageCond
    quality_order  
]

numeric_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

ordinal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(categories=ordinal_categories,
                               handle_unknown='use_encoded_value',
                               unknown_value=-1))
])

nominal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer([
    ('num', numeric_pipeline, numeric_cols),
    ('ord', ordinal_pipeline, ordinal_cols),
    ('nom', nominal_pipeline, nominal_cols)
])
lr_pca = Pipeline([
    ('preprocessor', preprocessor),
    ('pca', PCA(n_components=0.95)),
    ('regressor', GammaRegressor(max_iter=10000))
    #('regressor', LinearRegression())
])

lr_pca.fit(X_train, y_train)
y_pred_pca_train= lr_pca.predict(X_train)
y_pred_pca_test = lr_pca.predict(X_test)
rmse_train = math.sqrt(mean_squared_error(y_train, y_pred_pca_train))
rmse_test = math.sqrt(mean_squared_error(y_test, y_pred_pca_test))
print(f'Training set RMSE: {rmse_train}')
print(f'Test set RMSE: {rmse_test}')



Training set RMSE: 45213.09214634333
Test set RMSE: 31482.68259265181


In [14]:
gap = rmse_train - rmse_test
print("RMSE gap:", gap)
print("Gap ratio:", rmse_test / rmse_train)

RMSE gap: 13730.409553691523
Gap ratio: 0.6963178384426868


In [15]:
test_df = pd.read_csv('test.csv')
X_real_test = pd.get_dummies(test_df, drop_first=True)
X_real_test = X_real_test.reindex(columns=X.columns, fill_value=0)
X_real_test
test_predictions = lr_pca.predict(X_real_test)
submission = pd.DataFrame({
    "Id": X_real_test.Id,
    "SalePrice": test_predictions
})

submission.to_csv("predictions.csv", index=False)

In [16]:
## Random Forest Regressor 
## Gridsearch CV 
import pandas as pd
import math
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression, GammaRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, f1_score, recall_score, mean_squared_error
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV


X = df.drop('SalePrice', axis=1)
X['Electrical'] = X['Electrical'].bfill()
y = df['SalePrice']
# Define feature groups - numeric, ordinal and nominal(categorical)

numeric_cols = X.select_dtypes(include='number').columns.tolist()

ordinal_cols = [
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'HeatingQC',
    'KitchenQual',
    'FireplaceQu',
    'GarageQual',
    'GarageCond'
]

categorical_cols = X.select_dtypes(include=['object', 'string', 'category']).columns.tolist()
nominal_cols = [col for col in categorical_cols if col not in ordinal_cols]

# 2. Define the order for each ordinal column
quality_order = ['Po', 'Fa', 'TA', 'Gd', 'Ex']

ordinal_categories = [
    # ExterQual
    quality_order, 
    # ExterCond
    quality_order, 
    # BsmtQual
    quality_order,
    # BsmtCond
    quality_order,
    # HeatingQC
    quality_order,
    # KitchenQual
    quality_order,
    # FireplaceQu
    quality_order, 
    # GarageQual
    quality_order,
    # GarageCond
    quality_order  
]

numeric_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

ordinal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(categories=ordinal_categories,
                               handle_unknown='use_encoded_value',
                               unknown_value=-1))
])

nominal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer([
    ('num', numeric_pipeline, numeric_cols),
    ('ord', ordinal_pipeline, ordinal_cols),
    ('nom', nominal_pipeline, nominal_cols)
])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Full pipeline
rf_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("pca", PCA(n_components=0.95)),
    ("model", RandomForestRegressor(
        n_estimators=200,
        max_depth=None,
        random_state=42,
        n_jobs=-1
    ))
])

params_grid = {
    "pca__n_components": [0.90, 0.95, 0.99],
    "model__n_estimators": [100, 200],
    "model__max_depth": [None, 10, 15, 20, 25],
    "model__min_samples_split": [2, 5, 7], 
    "model__min_samples_leaf": [1, 2]
}

gs = GridSearchCV(estimator=rf_pipeline, param_grid=params_grid, cv=5, scoring="neg_root_mean_squared_error", n_jobs = 2, verbose=1)

gs.fit(X_train, y_train)
print("Best params:", gs.best_params_)
print("Best CV score:", gs.best_score_)

best_model = gs.best_estimator_
#print(best_model)

y_pred_pca_train= best_model.predict(X_train)
y_pred_pca_test = best_model.predict(X_test)
rmse_train = math.sqrt(mean_squared_error(y_train, y_pred_pca_train))
rmse_test = math.sqrt(mean_squared_error(y_test, y_pred_pca_test))
print(f'Training set RMSE: {rmse_train}')
print(f'Test set RMSE: {rmse_test}')

Fitting 5 folds for each of 180 candidates, totalling 900 fits
Best params: {'model__max_depth': 20, 'model__min_samples_leaf': 2, 'model__min_samples_split': 5, 'model__n_estimators': 100, 'pca__n_components': 0.9}
Best CV score: -29761.56593701615
Training set RMSE: 13532.853132962011
Test set RMSE: 32705.692272432338


Fitting 5 folds for each of 180 candidates, totalling 900 fits

Best params: {'model__max_depth': 20, 'model__min_samples_leaf': 2, 'model__min_samples_split': 5, 'model__n_estimators': 100, 'pca__n_components': 0.9}

Best CV score: -29761.56593701615

Training set RMSE: 13532.853132962011

Test set RMSE: 32705.692272432338

In [17]:
test_df = pd.read_csv('test.csv')

In [18]:
test_df

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,ScreenPorch,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition
0,1461,20,RH,80.0,11622,Pave,NaN,Reg,Lvl,AllPub,...,120,0,NaN,MnPrv,NaN,0,6,2010,WD,Normal
1,1462,20,RL,81.0,14267,Pave,NaN,IR1,Lvl,AllPub,...,0,0,NaN,NaN,Gar2,12500,6,2010,WD,Normal
2,1463,60,RL,74.0,13830,Pave,NaN,IR1,Lvl,AllPub,...,0,0,NaN,MnPrv,NaN,0,3,2010,WD,Normal
3,1464,60,RL,78.0,9978,Pave,NaN,IR1,Lvl,AllPub,...,0,0,NaN,NaN,NaN,0,6,2010,WD,Normal
4,1465,120,RL,43.0,5005,Pave,NaN,IR1,HLS,AllPub,...,144,0,NaN,NaN,NaN,0,1,2010,WD,Normal
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1454,2915,160,RM,21.0,1936,Pave,NaN,Reg,Lvl,AllPub,...,0,0,NaN,NaN,NaN,0,6,2006,WD,Normal
1455,2916,160,RM,21.0,1894,Pave,NaN,Reg,Lvl,AllPub,...,0,0,NaN,NaN,NaN,0,4,2006,WD,Abnorml
1456,2917,20,RL,160.0,20000,Pave,NaN,Reg,Lvl,AllPub,...,0,0,NaN,NaN,NaN,0,9,2006,WD,Abnorml
1457,2918,85,RL,62.0,10441,Pave,NaN,Reg,Lvl,AllPub,...,0,0,NaN,MnPrv,Shed,700,7,2006,WD,Normal


In [19]:
X_real_test = pd.get_dummies(test_df, drop_first=True)
X_real_test = X_real_test.reindex(columns=X.columns, fill_value=0)
X_real_test

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,LotShape,LandContour,Utilities,LotConfig,...,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,MiscVal,MoSold,YrSold,SaleType,SaleCondition
0,1461,20,0,80.0,11622,0,0,0,0,0,...,0,0,0,120,0,0,6,2010,0,0
1,1462,20,0,81.0,14267,0,0,0,0,0,...,36,0,0,0,0,12500,6,2010,0,0
2,1463,60,0,74.0,13830,0,0,0,0,0,...,34,0,0,0,0,0,3,2010,0,0
3,1464,60,0,78.0,9978,0,0,0,0,0,...,36,0,0,0,0,0,6,2010,0,0
4,1465,120,0,43.0,5005,0,0,0,0,0,...,82,0,0,144,0,0,1,2010,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1454,2915,160,0,21.0,1936,0,0,0,0,0,...,0,0,0,0,0,0,6,2006,0,0
1455,2916,160,0,21.0,1894,0,0,0,0,0,...,24,0,0,0,0,0,4,2006,0,0
1456,2917,20,0,160.0,20000,0,0,0,0,0,...,0,0,0,0,0,0,9,2006,0,0
1457,2918,85,0,62.0,10441,0,0,0,0,0,...,32,0,0,0,0,700,7,2006,0,0


In [20]:
#test_predictions = lr_pca.predict(X_real_test)
test_predictions = best_model.predict(X_real_test)
submission = pd.DataFrame({
    "Id": X_real_test.Id,
    "SalePrice": test_predictions
})

submission.to_csv("rf_predictions.csv", index=False)

In [21]:
import pandas as pd
import math
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.decomposition import PCA
from sklearn.linear_model import SGDRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, f1_score, recall_score, mean_squared_error
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import GridSearchCV


X = df.drop('SalePrice', axis=1)
X['Electrical'] = X['Electrical'].bfill()
y = df['SalePrice']
# Define feature groups - numeric, ordinal and nominal(categorical)

numeric_cols = X.select_dtypes(include='number').columns.tolist()

ordinal_cols = [
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'HeatingQC',
    'KitchenQual',
    'FireplaceQu',
    'GarageQual',
    'GarageCond'
]

categorical_cols = X.select_dtypes(include=['object', 'string', 'category']).columns.tolist()
nominal_cols = [col for col in categorical_cols if col not in ordinal_cols]

# 2. Define the order for each ordinal column
quality_order = ['Po', 'Fa', 'TA', 'Gd', 'Ex']

ordinal_categories = [
    # ExterQual
    quality_order, 
    # ExterCond
    quality_order, 
    # BsmtQual
    quality_order,
    # BsmtCond
    quality_order,
    # HeatingQC
    quality_order,
    # KitchenQual
    quality_order,
    # FireplaceQu
    quality_order, 
    # GarageQual
    quality_order,
    # GarageCond
    quality_order  
]

numeric_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

ordinal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(categories=ordinal_categories,
                               handle_unknown='use_encoded_value',
                               unknown_value=-1))
])

nominal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer([
    ('num', numeric_pipeline, numeric_cols),
    ('ord', ordinal_pipeline, ordinal_cols),
    ('nom', nominal_pipeline, nominal_cols)
])

# Full pipeline
sgd_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("pca", PCA(n_components=0.95)),
    ("model", SGDRegressor(random_state=42))
])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

params_grid = {
    "pca__n_components": [0.90, 0.95, 0.99],
    "model__penalty": ["l2", "elasticnet", "l1"],
    "model__alpha": [1e-5, 1e-4, 1e-3, 1e-2],
    "model__l1_ratio": [0.15, 0.5, 0.85],
    "model__max_iter": [1000,2000]
}

gs = GridSearchCV(estimator=sgd_pipeline, param_grid=params_grid, cv=5, scoring="neg_root_mean_squared_error", n_jobs = -1, verbose=1)

gs.fit(X_train, y_train)
print("Best params:", gs.best_params_)
print("Best CV score:", gs.best_score_)

best_model = gs.best_estimator_
#print(best_model)

y_pred_pca_train= best_model.predict(X_train)
y_pred_pca_test = best_model.predict(X_test)
rmse_train = math.sqrt(mean_squared_error(y_train, y_pred_pca_train))
rmse_test = math.sqrt(mean_squared_error(y_test, y_pred_pca_test))
print(f'Training set RMSE: {rmse_train}')
print(f'Test set RMSE: {rmse_test}')

Fitting 5 folds for each of 216 candidates, totalling 1080 fits
Best params: {'model__alpha': 0.01, 'model__l1_ratio': 0.15, 'model__max_iter': 1000, 'model__penalty': 'l2', 'pca__n_components': 0.99}
Best CV score: -34196.34396830431
Training set RMSE: 30173.119937821375
Test set RMSE: 36092.00760974403


Fitting 5 folds for each of 216 candidates, totalling 1080 fits

Best params: {'model__alpha': 0.01, 'model__l1_ratio': 0.15, 'model__max_iter': 1000, 'model__penalty': 'l2', 'pca__n_components': 0.99}

Best CV score: -34196.34396830431

Training set RMSE: 30173.119937821375

Test set RMSE: 36092.00760974403

In [22]:
test_df = pd.read_csv('test.csv')
X_real_test = pd.get_dummies(test_df, drop_first=True)
X_real_test = X_real_test.reindex(columns=X.columns, fill_value=0)
X_real_test
test_predictions = best_model.predict(X_real_test)
submission = pd.DataFrame({
    "Id": X_real_test.Id,
    "SalePrice": test_predictions
})

submission.to_csv("sgd_predictions.csv", index=False)

In [25]:
%pip install xgboost

Defaulting to user installation because normal site-packages is not writeable
   ---------------------------------------- 0.0/101.7 MB ? eta -:--:--
   -- ------------------------------------- 5.8/101.7 MB 29.4 MB/s eta 0:00:04
   ----- ---------------------------------- 13.4/101.7 MB 33.6 MB/s eta 0:00:03
   -------- ------------------------------- 21.2/101.7 MB 34.4 MB/s eta 0:00:03
   ---------- ----------------------------- 27.3/101.7 MB 33.3 MB/s eta 0:00:03
   -------------- ------------------------- 36.7/101.7 MB 35.9 MB/s eta 0:00:02
   ----------------- ---------------------- 44.6/101.7 MB 35.4 MB/s eta 0:00:02
   -------------------- ------------------- 52.7/101.7 MB 36.1 MB/s eta 0:00:02
   ------------------------ --------------- 61.6/101.7 MB 36.7 MB/s eta 0:00:02
   --------------------------- ------------ 69.5/101.7 MB 36.9 MB/s eta 0:00:01
   ------------------------------ --------- 78.1/101.7 MB 37.5 MB/s eta 0:00:01
   ---------------------------------- ----- 86.8/101


[notice] A new release of pip is available: 25.0.1 -> 26.0.1
[notice] To update, run: C:\Users\melvi\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [29]:
import pandas as pd
import math
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.decomposition import PCA
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, f1_score, recall_score, mean_squared_error
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import GridSearchCV


X = df.drop('SalePrice', axis=1)
X['Electrical'] = X['Electrical'].bfill()
y = df['SalePrice']
# Define feature groups - numeric, ordinal and nominal(categorical)

numeric_cols = X.select_dtypes(include='number').columns.tolist()

ordinal_cols = [
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'HeatingQC',
    'KitchenQual',
    'FireplaceQu',
    'GarageQual',
    'GarageCond'
]

categorical_cols = X.select_dtypes(include=['object', 'string', 'category']).columns.tolist()
nominal_cols = [col for col in categorical_cols if col not in ordinal_cols]

# 2. Define the order for each ordinal column
quality_order = ['Po', 'Fa', 'TA', 'Gd', 'Ex']

ordinal_categories = [
    # ExterQual
    quality_order, 
    # ExterCond
    quality_order, 
    # BsmtQual
    quality_order,
    # BsmtCond
    quality_order,
    # HeatingQC
    quality_order,
    # KitchenQual
    quality_order,
    # FireplaceQu
    quality_order, 
    # GarageQual
    quality_order,
    # GarageCond
    quality_order  
]

numeric_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

ordinal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(categories=ordinal_categories,
                               handle_unknown='use_encoded_value',
                               unknown_value=-1))
])

nominal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer([
    ('num', numeric_pipeline, numeric_cols),
    ('ord', ordinal_pipeline, ordinal_cols),
    ('nom', nominal_pipeline, nominal_cols)
])

# Full pipeline
xgb_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model", XGBRegressor(
        objective="reg:squarederror",
        random_state=42,
        n_jobs=-1
    ))
])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

params_grid = {
    #"pca__n_components": [0.90, 0.95, 0.99],
    "model__n_estimators": [200, 500],
    "model__max_depth": [3, 5, 7],
    "model__learning_rate": [0.03, 0.05, 0.1],
    "model__subsample": [0.8, 1.0],
    "model__colsample_bytree": [0.8, 1.0],
    "model__min_child_weight": [1, 3, 5]
}

gs = GridSearchCV(estimator=xgb_pipeline, param_grid=params_grid, cv=5, scoring="neg_root_mean_squared_error", n_jobs = -1, verbose=1)

gs.fit(X_train, y_train)
print("Best params:", gs.best_params_)
print("Best CV score:", gs.best_score_)

best_model = gs.best_estimator_
#print(best_model)

y_pred_no_pca_train= best_model.predict(X_train)
y_pred_no_pca_test = best_model.predict(X_test)
rmse_train = math.sqrt(mean_squared_error(y_train, y_pred_no_pca_train))
rmse_test = math.sqrt(mean_squared_error(y_test, y_pred_no_pca_test))
print(f'Training set RMSE: {rmse_train}')
print(f'Test set RMSE: {rmse_test}')

Fitting 5 folds for each of 216 candidates, totalling 1080 fits
Best params: {'model__colsample_bytree': 0.8, 'model__learning_rate': 0.05, 'model__max_depth': 3, 'model__min_child_weight': 1, 'model__n_estimators': 500, 'model__subsample': 0.8}
Best CV score: -27217.357421875
Training set RMSE: 9379.384201534767
Test set RMSE: 24866.092576036146


Fitting 5 folds for each of 216 candidates, totalling 1080 fits

Best params: {'model__colsample_bytree': 0.8, 'model__learning_rate': 0.05, 'model__max_depth': 3, 'model__min_child_weight': 1, 'model__n_estimators': 500, 'model__subsample': 0.8}

Best CV score: -27217.357421875

Training set RMSE: 9379.384201534767

Test set RMSE: 24866.092576036146

In [30]:
test_df = pd.read_csv('test.csv')
X_real_test = pd.get_dummies(test_df, drop_first=True)
X_real_test = X_real_test.reindex(columns=X.columns, fill_value=0)
X_real_test
test_predictions = best_model.predict(X_real_test)
submission = pd.DataFrame({
    "Id": X_real_test.Id,
    "SalePrice": test_predictions
})

submission.to_csv("xgb_no_pca_predictions.csv", index=False)

In [31]:
import pandas as pd
import math
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.decomposition import PCA
from xgboost import XGBRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, f1_score, recall_score, mean_squared_error
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import GridSearchCV


X = df.drop('SalePrice', axis=1)
X['Electrical'] = X['Electrical'].bfill()
y = df['SalePrice']
# Define feature groups - numeric, ordinal and nominal(categorical)

numeric_cols = X.select_dtypes(include='number').columns.tolist()

ordinal_cols = [
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'HeatingQC',
    'KitchenQual',
    'FireplaceQu',
    'GarageQual',
    'GarageCond'
]

categorical_cols = X.select_dtypes(include=['object', 'string', 'category']).columns.tolist()
nominal_cols = [col for col in categorical_cols if col not in ordinal_cols]

# 2. Define the order for each ordinal column
quality_order = ['Po', 'Fa', 'TA', 'Gd', 'Ex']

ordinal_categories = [
    # ExterQual
    quality_order, 
    # ExterCond
    quality_order, 
    # BsmtQual
    quality_order,
    # BsmtCond
    quality_order,
    # HeatingQC
    quality_order,
    # KitchenQual
    quality_order,
    # FireplaceQu
    quality_order, 
    # GarageQual
    quality_order,
    # GarageCond
    quality_order  
]

numeric_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

ordinal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(categories=ordinal_categories,
                               handle_unknown='use_encoded_value',
                               unknown_value=-1))
])

nominal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer([
    ('num', numeric_pipeline, numeric_cols),
    ('ord', ordinal_pipeline, ordinal_cols),
    ('nom', nominal_pipeline, nominal_cols)
])

# Full pipeline
xgb_pipeline = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("pca", PCA(n_components=0.95)),
    ("model", XGBRegressor(
        objective="reg:squarederror",
        random_state=42,
        n_jobs=-1
    ))
])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

params_grid = {
    "pca__n_components": [0.90, 0.95, 0.99],
    "model__n_estimators": [200, 500],
    "model__max_depth": [3, 5, 7],
    "model__learning_rate": [0.03, 0.05, 0.1],
    "model__subsample": [0.8, 1.0],
    "model__colsample_bytree": [0.8, 1.0],
    "model__min_child_weight": [1, 3, 5]
}

gs = GridSearchCV(estimator=xgb_pipeline, param_grid=params_grid, cv=5, scoring="neg_root_mean_squared_error", n_jobs = -1, verbose=1)

gs.fit(X_train, y_train)
print("Best params:", gs.best_params_)
print("Best CV score:", gs.best_score_)

best_model = gs.best_estimator_
#print(best_model)

y_pred_pca_train= best_model.predict(X_train)
y_pred_pca_test = best_model.predict(X_test)
rmse_train = math.sqrt(mean_squared_error(y_train, y_pred_pca_train))
rmse_test = math.sqrt(mean_squared_error(y_test, y_pred_pca_test))
print(f'Training set RMSE: {rmse_train}')
print(f'Test set RMSE: {rmse_test}')

Fitting 5 folds for each of 648 candidates, totalling 3240 fits
Best params: {'model__colsample_bytree': 1.0, 'model__learning_rate': 0.1, 'model__max_depth': 3, 'model__min_child_weight': 1, 'model__n_estimators': 500, 'model__subsample': 0.8, 'pca__n_components': 0.99}
Best CV score: -27508.10546875
Training set RMSE: 2262.521491610632
Test set RMSE: 31209.0356147062


Fitting 5 folds for each of 648 candidates, totalling 3240 fits

Best params: {'model__colsample_bytree': 1.0, 'model__learning_rate': 0.1, 'model__max_depth': 3, 'model__min_child_weight': 1, 'model__n_estimators': 500, 'model__subsample': 0.8, 'pca__n_components': 0.99}

Best CV score: -27508.10546875

Training set RMSE: 2262.521491610632

Test set RMSE: 31209.0356147062

In [32]:
test_df = pd.read_csv('test.csv')
X_real_test = pd.get_dummies(test_df, drop_first=True)
X_real_test = X_real_test.reindex(columns=X.columns, fill_value=0)
X_real_test
test_predictions = best_model.predict(X_real_test)
submission = pd.DataFrame({
    "Id": X_real_test.Id,
    "SalePrice": test_predictions
})

submission.to_csv("xgb_pca_predictions.csv", index=False)